core/src/main/java/io/questdb/cairo/sql/Function.java

```diff
@@ -172,13 +172,26 @@ default boolean isOperator() {
         return false;
     }
 
+    /**
+     * Returns true if the function supports parallel execution, e.g. parallel filter
+     * or GROUP BY. If the method returns false, single-threaded execution plan
+     * must be chosen for the query.
+     *
+     * @return true if the function and all of its children functions support parallel execution
+     */
+    default boolean isParallelismSupported() {
```

Comment:
We need both flags. That's because some parallel-friendly aggregate and "normal" functions are thread-safe and some aren't. In the latter case, we don't need to clone them per worker thread. I'd rename `isReadThreadSafe` to `isThreadSafe` since in the case of aggregate functions it also means write thread-safety, but then the patch would become even bigger and less reviewable.

Reply:
nit: it would be good to explain in the comments the implications of having a parallelizable function that is not thread-safe with an example